### PR TITLE
NSHMDB Improvements for Workflow + Rupture Explorer

### PR DESCRIPTION
**Pull request by lispandfound**
This pull request includes several changes to the repository, mostly in service of the new workflow and the rupture explorer. Here are a summary of the changes:

# Performance Improvements
- The nshmdb generator directly exports the pandas dataframes containing the rupture <--> fault segment join tables directly to the database using `DataFrame.to_sql`. This results in the database generation taking less than a minute instead of the 15 minutes it used to take.
- A handful of SQL query optimisations and indexes for some searches, and the introduction of [duckdb](https://duckdb.org/) to query the database for the most complex aggregation queries.

# Database Contents
I added a number of properties from the old CRU solutions zip into the database including:

- [Magnitude frequency distributions](https://hazard.openquake.org/gem/methods/mfd/) for every segment of every fault in the database. These are used to determine the most likely initial fault in a rupture for the workflow. You can see a visual of this by clicking one of the rupture map buttons on the rupture explorer (e.g. for [this search](https://quakecoresoft.canterbury.ac.nz/ruptures/?query=Alpine:+Caswell+%26+Alpine:+Caswell+-+South+George+%26+(Alpine:+George+seaward+|+Alpine:+George+landward)+%26+Alpine:+George+to+Jacksons+%26+Alpine:+Jacksons+to+Kaniere&rate_lower_bound=-4.5)).
image

- Rupture properties including: magnitude, area, length, and rupture rate extracted from the database.
- Fault info not directly related to geometry including: name, rake, and tectonic type.

# Database querying

Introduced a number of useful querying features, most notably: 

- `NSHMDB.query` which is a powerful method to query the database for complex expressions. For example:
```python
# Query the database for all ruptures containing the spine of the Alpine fault:
db = nshmdb.NSHMDB(path_to_database)
    ruptures = db.query('Alpine: Caswell & Alpine: Caswell - South George & (Alpine: George seaward | Alpine: George landward) & Alpine: George to Jacksons & Alpine: Jacksons to Kaniere',
    rate_bounds = (1e-5, None)
)
``` 
This required a custom parser to convert the search queries into SQL, which is the contents of the `nshmdb.query` module.

### CODE REVIEW

**Review by joelridden — DISMISSED** [review]
Seems good, didnt go into depth with the SQl checking though or stuff about duckDB in the query.py as I'm unsure what some of those functions are achieving overall.

**Comment by lispandfound** [comment]
After discussions with @AndrewRidden-Harper, I think the parsing part of the query module is likely to be extracted out of the repository (perhaps into qcore?) to allow it's reuse in other related projects. To address what @joelridden is asking about the point of the module is, the aim of the query module is to allow the user to make sophisticated queries on the NSHM database based on the inclusion of faults. 

# Why Did I Create the Query Module

To understand this you need to understand why I made the rupture explorer. The motivating reason to create our own rupture explorer is the lack of specificity in the GNS developed NSHM rupture explorer. That browser lets you select say 10 faults and then it will find all ruptures containing *at least one* of those ten faults. This make it really hard to select ruptures containing *all of a list of faults*.

It actually gets worse than this, Brendon wanted to find ruptures in the database that contained the base of the alpine fault. These are ruptures that contain:

1. Alpine: Caswell
2. Alpine: Caswell - South George
3. Alpine: George seaward *or* Alpine: George landward, then
4. Alpine: George to Jacksons
5. Alpine: Jacksons to Kaniere

This kind of querying is impossible with the GNS rupture explorer, and also impossible with Sung's redevelopment of the rupture explorer because it lacks the ability to express both logical and, and logical or expressions in the same query. The simplest solution I believe is a simple query language that lets you write the above as:

```
Alpine: Caswell & Alpine: Caswell - South George & (Alpine: George seaward | Alpine: George landward) & Alpine: George to Jacksons & Alpine: Jacksons to Kaniere
```

This statement needs to be first parsed, and then compiled to an SQL expression to search the database. So the query module was developed to power the complex search queries of the rupture explorer.

# The Design of the Query Module 

The query module has three parts: a lexer, a parser and `to_sql` to compile the parsed expression to an SQL expression. 
The lexer turns a query string like 

```
Alpine: Caswell & Alpine: Caswell - South George & (Alpine: George seaward | Alpine: George landward) & Alpine: George to Jacksons & Alpine: Jacksons to Kaniere
```

Into a stream of tokens like

```
[Token(token_type=<TokenType.FAULT: 5>, value='Alpine: Caswell'), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.AND: (3, 4)>), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: Caswell - South George'), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.AND: (3, 4)>), Token(token_type=<TokenType.LPAR: 1>, value=None), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: George seaward'), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.OR: (1, 2)>), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: George landward'), Token(token_type=<TokenType.RPAR: 2>, value=None), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.AND: (3, 4)>), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: George to Jacksons'), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.AND: (3, 4)>), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: Jacksons to Kaniere')]
```

Then the job of the parser is to convert that stream of tokens into a tree that represents the order of operations:

```
┐
└── InfixOperator.AND
    ├── 0
    │   └── InfixOperator.AND
    │       ├── 0
    │       │   └── InfixOperator.AND
    │       │       ├── 0
    │       │       │   └── InfixOperator.AND
    │       │       │       ├── 0: Alpine: Caswell
    │       │       │       └── 1: Alpine: Caswell - South George
    │       │       └── 1
    │       │           └── InfixOperator.OR
    │       │               ├── 0: Alpine: George seaward
    │       │               └── 1: Alpine: George landward
    │       └── 1: Alpine: George to Jacksons
    └── 1: Alpine: Jacksons to Kaniere
```

The `to_sql` function takes a query (and a few other parameters), and compiles this into SQL that can search the database.

```
('SELECT\n'
 '     rupture.rupture_id, ANY_VALUE(rupture.magnitude), '
 'ANY_VALUE(rupture.area), ANY_VALUE(rupture.len), ANY_VALUE(rupture.rate)\n'
 '    FROM rupture\n'
 '    JOIN\n'
 '        rupture_faults ON rupture.rupture_id = rupture_faults.rupture_id\n'
 '    JOIN\n'
 '        fault ON rupture_faults.fault_id = fault.fault_id\n'
 '    JOIN\n'
 '        parent_fault ON fault.parent_id = parent_fault.parent_id\n'
 '    WHERE rupture.rate IS NOT NULL  \n'
 '    GROUP BY rupture.rupture_id\n'
 '    HAVING  (((((SUM(CASE WHEN parent_fault.name = ? THEN 1 ELSE 0 END) > 0) '
 'AND (SUM(CASE WHEN parent_fault.name = ? THEN 1 ELSE 0 END) > 0)) AND '
 '((SUM(CASE WHEN parent_fault.name = ? THEN 1 ELSE 0 END) > 0) OR (SUM(CASE '
 'WHEN parent_fault.name = ? THEN 1 ELSE 0 END) > 0))) AND (SUM(CASE WHEN '
 'parent_fault.name = ? THEN 1 ELSE 0 END) > 0)) AND (SUM(CASE WHEN '
 'parent_fault.name = ? THEN 1 ELSE 0 END) > 0))\n'
 '    ORDER BY ANY_VALUE(rupture.rate)\n'
 '    DESC NULLS LAST\n'
 '    LIMIT ?\n'
 '    ',
 ['Alpine: Caswell',
  'Alpine: Caswell - South George',
  'Alpine: George seaward',
  'Alpine: George landward',
  'Alpine: George to Jacksons',
  'Alpine: Jacksons to Kaniere',
  100])
```
# Where I Think This Module Could be Useful

It's a bit hard to explain the vision for this. I plan to make a demo out of @AndrewRidden-Harper's CPT data that shows how we can make this query module really reusable to allow us to quickly  produce dashboards powered on the backend by pandas dataframes (think: CPT data, NSHMDB data, GCMT data, IMs) integrated with ChatGPT to allow powerful natural language queries like:

```
All large ruptures containing the faults in the alpine fault system.
```
or
```
All high-quality CPT records in the south island with a depth no greater than 30m. 
```

**Comment by joelridden** [comment]
> After discussions with @AndrewRidden-Harper, I think the parsing part of the query module is likely to be extracted out of the repository (perhaps into qcore?) to allow it's reuse in other related projects. To address what @joelridden is asking about the point of the module is, the aim of the query module is to allow the user to make sophisticated queries on the NSHM database based on the inclusion of faults.
> # Why Did I Create the Query Module
> 
> To understand this you need to understand why I made the rupture explorer. The motivating reason to create our own rupture explorer is the lack of specificity in the GNS developed NSHM rupture explorer. That browser lets you select say 10 faults and then it will find all ruptures containing _at least one_ of those ten faults. This make it really hard to select ruptures containing _all of a list of faults_.
> 
> It actually gets worse than this, Brendon wanted to find ruptures in the database that contained the base of the alpine fault. These are ruptures that contain:
> 
>     1. Alpine: Caswell
> 
>     2. Alpine: Caswell - South George
> 
>     3. Alpine: George seaward _or_ Alpine: George landward, then
> 
>     4. Alpine: George to Jacksons
> 
>     5. Alpine: Jacksons to Kaniere
> 
> 
> This kind of querying is impossible with the GNS rupture explorer, and also impossible with Sung's redevelopment of the rupture explorer because it lacks the ability to express both logical and, and logical or expressions in the same query. The simplest solution I believe is a simple query language that lets you write the above as:
> 
> ```
> Alpine: Caswell & Alpine: Caswell - South George & (Alpine: George seaward | Alpine: George landward) & Alpine: George to Jacksons & Alpine: Jacksons to Kaniere
> ```
> 
> This statement needs to be first parsed, and then compiled to an SQL expression to search the database. So the query module was developed to power the complex search queries of the rupture explorer.
> # The Design of the Query Module
> 
> The query module has three parts: a lexer, a parser and `to_sql` to compile the parsed expression to an SQL expression. The lexer turns a query string like
> 
> ```
> Alpine: Caswell & Alpine: Caswell - South George & (Alpine: George seaward | Alpine: George landward) & Alpine: George to Jacksons & Alpine: Jacksons to Kaniere
> ```
> 
> Into a stream of tokens like
> 
> ```
> [Token(token_type=<TokenType.FAULT: 5>, value='Alpine: Caswell'), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.AND: (3, 4)>), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: Caswell - South George'), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.AND: (3, 4)>), Token(token_type=<TokenType.LPAR: 1>, value=None), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: George seaward'), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.OR: (1, 2)>), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: George landward'), Token(token_type=<TokenType.RPAR: 2>, value=None), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.AND: (3, 4)>), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: George to Jacksons'), Token(token_type=<TokenType.UNARY_OPERATOR: 3>, value=<InfixOperator.AND: (3, 4)>), Token(token_type=<TokenType.FAULT: 5>, value='Alpine: Jacksons to Kaniere')]
> ```
> 
> Then the job of the parser is to convert that stream of tokens into a tree that represents the order of operations:
> 
> ```
> ┐
> └── InfixOperator.AND
>     ├── 0
>     │   └── InfixOperator.AND
>     │       ├── 0
>     │       │   └── InfixOperator.AND
>     │       │       ├── 0
>     │       │       │   └── InfixOperator.AND
>     │       │       │       ├── 0: Alpine: Caswell
>     │       │       │       └── 1: Alpine: Caswell - South George
>     │       │       └── 1
>     │       │           └── InfixOperator.OR
>     │       │               ├── 0: Alpine: George seaward
>     │       │               └── 1: Alpine: George landward
>     │       └── 1: Alpine: George to Jacksons
>     └── 1: Alpine: Jacksons to Kaniere
> ```
> 
> The `to_sql` function takes a query (and a few other parameters), and compiles this into SQL that can search the database.
> 
> ```
> ('SELECT\n'
>  '     rupture.rupture_id, ANY_VALUE(rupture.magnitude), '
>  'ANY_VALUE(rupture.area), ANY_VALUE(rupture.len), ANY_VALUE(rupture.rate)\n'
>  '    FROM rupture\n'
>  '    JOIN\n'
>  '        rupture_faults ON rupture.rupture_id = rupture_faults.rupture_id\n'
>  '    JOIN\n'
>  '        fault ON rupture_faults.fault_id = fault.fault_id\n'
>  '    JOIN\n'
>  '        parent_fault ON fault.parent_id = parent_fault.parent_id\n'
>  '    WHERE rupture.rate IS NOT NULL  \n'
>  '    GROUP BY rupture.rupture_id\n'
>  '    HAVING  (((((SUM(CASE WHEN parent_fault.name = ? THEN 1 ELSE 0 END) > 0) '
>  'AND (SUM(CASE WHEN parent_fault.name = ? THEN 1 ELSE 0 END) > 0)) AND '
>  '((SUM(CASE WHEN parent_fault.name = ? THEN 1 ELSE 0 END) > 0) OR (SUM(CASE '
>  'WHEN parent_fault.name = ? THEN 1 ELSE 0 END) > 0))) AND (SUM(CASE WHEN '
>  'parent_fault.name = ? THEN 1 ELSE 0 END) > 0)) AND (SUM(CASE WHEN '
>  'parent_fault.name = ? THEN 1 ELSE 0 END) > 0))\n'
>  '    ORDER BY ANY_VALUE(rupture.rate)\n'
>  '    DESC NULLS LAST\n'
>  '    LIMIT ?\n'
>  '    ',
>  ['Alpine: Caswell',
>   'Alpine: Caswell - South George',
>   'Alpine: George seaward',
>   'Alpine: George landward',
>   'Alpine: George to Jacksons',
>   'Alpine: Jacksons to Kaniere',
>   100])
> ```
> 
> # Where I Think This Module Could be Useful
> 
> It's a bit hard to explain the vision for this. I plan to make a demo out of @AndrewRidden-Harper's CPT data that shows how we can make this query module really reusable to allow us to quickly produce dashboards powered on the backend by pandas dataframes (think: CPT data, NSHMDB data, GCMT data, IMs) integrated with ChatGPT to allow powerful natural language queries like:
> 
> ```
> All large ruptures containing the faults in the alpine fault system.
> ```
> 
> or
> 
> ```
> All high-quality CPT records in the south island with a depth no greater than 30m. 
> ```

Thank you for this! Great explanation and now I understand the vision behind why this is useful, especially turning plain English into a query. Not sure though if for example plain english transcribed through AI could just generate the query and then could just skip the middleman for all the logical ors and ands stuff, but maybe thats for future.

**Comment by lispandfound** [comment]
ChatGPT could possibly generate the tree? You wouldn't want it generating the search code though because I feel like that would be potentially buggy.

**Comment by lispandfound** [comment]
I just added a bunch of test cases (not sure about coverage). For some reason the pytest check is failing at the apt-get stage because it can't find `libcurl`. That seems like a GitHub bug, probably a caching issue, I'll try resub it tomorrow.

File: nshmdb/query.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
+"""
```
Comment (claudio525) [inline nshmdb/query.py:1]: Nice work! Did you consider existing packages for this?

Comment (lispandfound) [reply nshmdb/query.py:1]: Tried looking but couldn't find anything. There are really big expression parsing libraries (as in, made to build compilers with) that I could find but nothing as simple as "parse something that looks like mathematics and give me a tree".

File: tests/test_nshmdb.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    assert faults == {
+        "Alpine Fault": FaultInfo(
+            fault_id=1, name="Segment 1", parent_id=1, rake=90.0, tect_type=None
+        )
+    }
+
+
+def test_query(alpine_fault_nshmdb: NSHMDB):
+    ruptures = alpine_fault_nshmdb.query("Alpine Fault")
+    print(ruptures)
```
Comment (joelridden) [inline tests/test_nshmdb.py]: probably can go

**Review by joelridden — APPROVED** [review]

**Comment by claudio525** [comment]
> ChatGPT could possibly generate the tree? You wouldn't want it generating the search code though because I feel like that would be potentially buggy.

I think the important part will be ensuring that the LLM has the appropriate context. E.g. what are the valid columns in the dataframe, and in the case of non-numerical columns what are the potential unique values, etc., such that it can A) understand what you are asking and B) produce an as accurate as possible query or tree.
Very nice idea for LLM use case.

File: nshmdb/nshmdb.py [hunk 3]
```diff
@@ -21,20 +21,67 @@
...
+from qcore import coordinates
+from source_modelling.sources import Fault, Plane
 
-from nshmdb import fault
-from nshmdb.fault import Fault
+from nshmdb import query
 
 
 @dataclasses.dataclass
+class Rupture:
```
Comment (claudio525) [inline nshmdb/nshmdb.py:41]: Might be worth including the fault metdata as well? I.e. `FaultInfo` dict

Comment (lispandfound) [reply nshmdb/nshmdb.py:41]: you can get that from a separate call to get fault info so I don't think it's super necessary.

**Review by claudio525 — APPROVED** [review]

**Review by AndrewRidden-Harper — APPROVED** [review]
impressive

**Pull request merged**